src/ewah/bitset_encoder.zig

```diff
@@ -0,0 +1,234 @@
+                    source_words[source_index + 1..][0..literals_count]);
+                target_index += run_length + literals_count;
+                source_index += 1 + literals_count;
+            }
+            assert(target_index <= target.len);
+            assert(source_index == source_words.len);
+            return target_index;
+        }
+
+        // Returns the maximum number of bytes that `source` needs to encode to.
+        pub fn encode_size_max(source: []const Word) usize {
+            // Assume (pessimistically) that every word will be encoded as a literal.
+            const literals_count = source.len;
+            const markers = div_ceil(literals_count, marker_literals_max);
+            return (literals_count + markers) * @sizeOf(Word);
+        }
+
+        // Returns the number of bytes written to `target`.
+        pub fn encode(source: []const Word, target: []u8) usize {
+            assert(target.len == encode_size_max(source));
+
+            const target_words = @alignCast(@alignOf(Word), mem.bytesAsSlice(Word, target));
+            std.mem.set(Word, target_words, 0);
+            var target_index: usize = 0;
+            var source_index: usize = 0;
+            while (source_index < source.len) {
+                const word_next = source[source_index];
+                const run_length = rl: {
+                    if (is_literal(word_next)) break :rl 0;
+                    // Measure run length.
+                    const run_max = std.math.min(source.len, source_index + marker_run_length_max);
+                    var run_end: usize = source_index + 1;
+                    while (run_end < run_max and source[run_end] == word_next) : (run_end += 1) {}
+                    const run = run_end - source_index;
+                    source_index += run;
+                    break :rl run;
+                };
+                // For consistent encoding, set the run bit to 0 when there is no run.
+                const run_bit = if (run_length == 0) 0 else word_next & 1;
+                // Count sequential literals that immediately follow the run.
+                const literals_max = std.math.min(source.len - source_index, marker_literals_max);
+                const literal_count = for (source[source_index..][0..literals_max]) |word, i| {
+                    if (!is_literal(word)) break i;
+                } else literals_max;
+
+                target_words[target_index] = marker(.{
+                    .run_bit = @intCast(u1, run_bit),
+                    .run_length = @intCast(MarkerRunLength, run_length),
+                    .literals = @intCast(MarkerLiterals, literal_count),
+                });
+                std.mem.copy(Word, target_words[target_index + 1..][0..literal_count],
+                    source[source_index..][0..literal_count]);
+                target_index += 1 + literal_count; // +1 for the marker word
```

Comment:
Since there's a comment explaining the `1 + ` and the lines are very near together anyway I think it's ok as-is. If it was more spread out I'd split it into two increments.